### DISCUSSION TITLE
Typescript errors are undebuggable due to super generic / inference objects

### DISCUSSION THREAD

**Discussion by Raynos** [wevm/viem#1301] [Question]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Package Version

1.10.14

### Current Behavior

When i get a typescript error related the viem client it prints an unreadably large object

![image](https://github.com/wagmi-dev/viem/assets/479538/74a23590-5a2a-4450-8540-397b4a0ca8e0)

### Expected Behavior

Expect the types to be readable.

### Steps To Reproduce

I tried to use the types from the viem package like

```
import { createPublicClient, http } from 'viem';

// https://viem.sh/docs/clients/transports/http.html#retrycount-optional
// defaults include 3 retries, retry delay of 150ms with exponential backoff, 10s timeout
function viemClientForChainHelper(chain: Chains): ViemClientForChain {
  const viemChain = chainToViemChain(chain);
  return createPublicClient({
    chain: viemChain,
    transport: http(rpcForChainId(viemChain.id), {
      // https://viem.sh/docs/clients/transports/http.html#batch-json-rpc
      batch: { wait: 10 },
    }),
    batch: {
      // https://viem.sh/docs/clients/public.html#batch-multicall-wait-optional
      multicall: { wait: 10 },
    },
  });
}


export type ViemClientForChain = PublicClient<
  Transport,
  Chain<ChainFormatters>
>;
```

and I get an error on `createPublicClient`

**Comment by Raynos** [comment]
When i tried to read and debug what what was going on I found out that the core implementation of viem is super generic and reflection based ( https://github.com/wagmi-dev/viem/blob/main/src/clients/createClient.ts#L146-L191 )

**Comment by Raynos** [comment]
Implementing the client as a Class and not using object literals / extend / object.assign would fix this.

it would also make the code readable.

**Reply by jxom** [reply]
Implementing as a Class will have the same effect on the type inferrence and emit the same error.

**Comment by tmm** [comment]
Can you create a [TypeScript Playground](https://www.typescriptlang.org/play) with a minimal reproducible example? Impossible to tell what's going wrong here (and if it's related to Viem) without the full picture.

**Comment by jxom** [comment]
Maybe this playground will help your case:

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbwFBzgYygUwIY0wBQFcAjAG2DQGFzMA7GAGhTgAsYYwnUYBPMTOJRbZgtLnF784RMhWrA6jZpIEAVKNloBnSLCYBfOADMoEEHABEANwUgLAbiShd8AFRxsW9MNFeTZyxtMEAB6NB9tByQkFWkScioaegAxaCERWgAecIy4TAAPPFoAEy900TgAHzhCEswjUUxiuABeQQiqmrqG2iaAPlbmGQT5RUz1TR1oRm8MvsckI1q0GGAIWnQsXAJ4uSSYVKhyrOYcioKi0vbc6tri+sbmtruH3uKmPoAKM9oALlnRABKf7DPYKFJpCLZCIDZCoLAwQhQDYYHB4UGJcEwT4IAFiCQabQuf5sDifQFwfSAxz6aJodZaeBofYAfQAjK1NmidrJMYpDsdvhEtAA6EAZXowQFIZlY9kin5IEIhVCq1AAPQA-HSGUzWQAmTmo7YY0YQo4RIUZUUQMCrEDALQgaWyxQs-UKiJKlVqjWaoA

**Comment by Raynos** [comment]
The playground example you gave says

```
Type instantiation is excessively deep and possibly infinite.
``` 

Let me see if I can make one.

**Reply by jxom** [reply]
Interesting, seems to work fine for me. Sometimes it takes a while to install modules, etc in the Playground - maybe that's why? 

<img width="966" alt="Screenshot 2023-10-06 at 9 16 27 am" src="https://github.com/wagmi-dev/viem/assets/7336481/bbb62a4c-b09d-4fc2-904f-7bc4253229b9">

**Comment by Raynos** [comment]
https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQBQK4BGANsMgMIloB2MANHABYwxj37GkXDV1wAqUDFQDOkWHAC+cAGZQIIOAHIAbtxCKA3AChQY+EgDmENFBL0QGYFSppeEMDFDBhIevcchnIAOLHTwSRk5BRU1AHpkBksRTS0tMLC4XWh9ODQADyi8YUdlNG88DCgAE0DZeSUAAWEIPCpi9IBPAC8wvEciYVitDL0UCBF4AEkaEyoMIjIoqzgAXkQtODgAWQBBIYA5DYBRPgAuJTXNnb5FWkW4bwB5bYAlABkhg8Vru8ezi6ucPiHloYBlZbPL4-P6Aj5LEG-AHLAD6rweTyUULBcIR73OUgwwn6g20vRScBgjTAaDgU2iOPmAApiaSINI4CNsFBxpNplQAJQAbQA1mhGgyiSS0ELmWMJhSrABdbRaaR1ZCOAYoDl8CAANTUUqo1Mi0QOOuEnIWS2EAHdgDBInA9RyTQgLktkNiyYojltdoo9k6lnB0DA8Ky4BYrDYYNo-ShXUp0UNvb6lgGg1Q4EYTCRI36XcI3SiYQmo0nbCm4O4nC4s86Y4p84D4TdEYWi8ng+XPC5fBngFmJFo+z10n1kAMcqrokNilTTSt1p79nAAIznJZxg4AVhXcDrQKXAAYtzuG28kQAWABMB-7cGxuLHwiwzmk3BxtzQI5KAB4jfQqHgQAQJgAHxynEBLiCOgxwMIeBgHoaDFDqk7TtyFzLnACQrNEtgXJuGGJF2-hoQe+Hbg4FYgBcF4kZhVzkR2CiESQWjSjeOKQTkcrgfAdJkv8sHwYhHKTnMtq8UKMFwSkCFIVOPJ-gBJiymBQ6EgqVBKsAKpQGAyAAGLQLJdoTsUBz8VJsAycJxScgcORQFYBgzkQthwEGRA4Ogz7pHZMAOVQBhymalrWgwtr6lYk4Or6OZkhFVDIQAdB6Jw+kWbmmJ5aDeaJii2AwAC0obWLYsTpQQ6AYLyVbRrm46RVOiVromGUeV5wDpLl+UFem-hlUWFWYNVMUxvFSU7mlRbuVlOXzIo7hFdE4b9VGg1VTVsX1QljVHs16XTe1nVzQtvUkCtfprcNSwDsWgbBgABkwLDCHsCQACQIAd2UdRIiUGIlEyRGgICNIlI4gGEyjnmEH2KIoEj3doA7yoqyqpqowNcDwBlQDqAASaBEKSUDGVYhocsaBzsCQ5CUDQn4CEIogpEBM4cfAGMgDqonxeqWqYxypNclmrapqgmDYNTnB0zA1KOlG8UHJzOpbksflM3oBxPWA1I6fphnWdSyscolwA2fQ8tFph2svQknOJcIDBhMUEDIMIEQy+76siHo7va4lTAgEQADEBBYJEBUAFY1FQBV6y1YehQcSDmpYMAHIue6SKrkicjnieRMnLXW8wYC25DagO07Ltux73A0O7YCEDTAcwEHofh4VIB4EQjgukQRAFanVoFeWAwTC13e96QExEMncDD+n+7Z76EhbhInJI3E7MoDLZAYEDok2OaWFgN+FNsM30v1zADOCD7LNAdSm8qX06maSqBi2PzXMyzjOpC3JpSWycApa0xvnfDWLNnKuWVjLHme8D4MDQH9WwQsX5LGAIyakABCOBN9opRnwTwUSxCaD-w5ATImJh0EbUQUDB2aD4r0DITADBkg4i3VLKwreg436oy0qmL+MAf5YxoAAdStAwAAQp3RyAARZwGBiAIWpBcRW5IL5aBAWAsRt9GYP1gKzS2O9jbRB5mqTU2pBbxRfhcUWKBKqSyvuAngcsYociVtY6IOdvbM1gFrUuutdIUJMkbbxVhTbmxnKtTuBxpATFzDnDe+c4kxL9FPPus94mJLQMk9eL8+xAA

**Comment by Raynos** [comment]
The typescript playground appears to cheat and uses `... 52 more...` which makes the typescript errors partially readable.

**Comment by Raynos** [comment]
The implementation of the viemclient uses a lot of generics and embedded generics and it's difficult to write static types for without relying on typescript return type inference or typescript generic inference.

Trying to use explicitly typed generics and being more exact is proving to be a challenge and then when something goes wrong it spits out an error about an enormous typescript inferred object literal instead of a particular field being wrong on a statically typed class.

**Reply by jxom** [reply]
Here is the Playground without errors: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgYygUwIYzQBQK4BGANsMgMIloB2MANHABYwxj37GkXDV1wAqUDFQDOkWPTIMMwKnAC+cAGZQIIOACIAbtxDqA3AChQY+EgDmENFBL0Q0qlTS8IYGKGDCQ9F25AeQAOKW1sDySipqWjoA9MhSMsL6BgbR0XDG0KZwaAAeUnjCbppoAXgYUAAmYcqqcADkAALCEHhUFTkAngBe0XhuRMJ1hga5JigQIvAAkjRWVBhEkvZwALyIBnBwALIAglMAcvsAonwAXBq7B8d86rQbcAEA8kcASgAyU+fqT68ft-ePHB8KZbKYAZS2X0BwNBEP+m2hIPBWwA+j93p8NIjYaj0X87goMMJxpNDKNMnAYB0wGg4EsEqs4AAKKk0iCKOAzbBQeaLeJUACUAG0ANZoDrsynUtCSrlzBb0qgAXWGyAmhRQ-L4EAAajpFYyEPdLocTuc7DJHHR7njMRYrDYAUCkRDzj53J47gjnTi0c8MW7XB7AsFHYTiWrSclybASRq4vYphViWsjZsTddzgBGL0Pf0fc4AVlz2OR2YADCWfci-b9MQAWABMlYM4bj8GEWA8im4xJeaDVlQAPIrhPQqHgQAQrAA+YYx+CRjXCPBgExoCqKpMpuBC+45uCpbb2Jz3YuHtJBB3AfeVi9wR5Bvyee5Nu9Hx++fwPUM3pVwIl23nHIxlZWkwVXddN35JNGRZaVJRXNdMg3Ldk2FCcpysFVoxAilFFaZA3AmOAoDAZAADFoDQpkExkJNzgg5DYFQmCKgFc5CigGQzHWTYiCcOA8GsHB0B7HJOJgbiqDMYZNmEAB3YAYDiZk6KoJMBT4zYUCJWl1O3AA6DMzXuHShJEsTgByRl1CcBgAFoLQcJwknMzYCHQDARUMczkD0zVE2TQzbVOMydOEohRLQcTbPshz7RCNz3M8zAfPC-zhH0tjhEM0tXXCzZIui2K1nUHwnPsK1kvM1LvN8nTMuyoLcvy3F80+QqLKiqybLKirEpIGqdLq9LNjkZJNnQGBhNkAADJgWGEU5UgAEgQYrerkQyzEMhY4jQEAOkMtUQGiTRG2idb1HUOQ5sMCaUlSZ6Xtet73o+z6vu+n6nrgCjCOIqhkgIqgiOAEjtEOrgeCoqBFQACTQIgaSgId1OyHJsDaYlRxnJl7nU851LuDi4HYEhyEoGghwEIRREyeh4LZDl1O1PVof5YV1KVGdtKXeAoZAA01jZ3V9X5IUeYa6bZpQLzsApzhqZgJk00a-lziFxVc02KT6ZMc5FrAJkyMo6i2KZbX+UM4B2PodXzKPY3ltSIXDOEBhogqCBkGEWIVf9-WRBMf3jcMpgQCIABiAgsDiByACtmioByza6uOVIYc4kAU6QYAreRdfkAVi8zuIc6653mDAV3zp0D2vZ9v2A+4Gh-bAQhKYjmAo9j+PHJAPAiDcfyiCIBy8+Uhz3QmBYuqHkfSAWIgc7gKeC7gLNyyL8K5FzOQBQe5IBZQFXGW1lW4cR5HUaZOoTL4OoBQMZAVZO-k-vMgA9AB+ZIfqAKAcAkBH0-oUCJMIE+RBIHky7srNuMAURv0gejfkmNsbJjpOggAPkJNoMUZAbkZAaPBrQKiEMcBUPm6siaBRkHJdsUA8BEWgGrehsg5BrzoRjQ+2k9YMA8B-ZYot+QNQmo9FBwhiQc2FlfC29hkEwOkUOQm6DciYNxvyMcBg+YaOoFgpWVNEFKNQczGUrMtTi05vYbm-Jeb83VFJFhMA2E8K5vw9scBrYiI4ezCW9gpb2IavJVcVh2F0J8TIEu9wJEnycWfExUidyOAUnAWRMMaDX35KY6R99H7P1firXJuV1Jf3cj-X+QA

You still need to do the same amount of generic slot passing to get inference to work correctly even for classes. Let me know if there is something I am missing.

**Reply by Raynos** [reply]
I believe making one thing a class won't help much, a bunch of things have to be made classes, made interfaces, have less higher order templates.

**Reply by jxom** [reply]
Can you fork my repro there and modify the class implementation to demonstrate what you mean? The class implementation in that playground is vanilla and has no external runtime dependencies, so should be straight forward to modify. It should not break the type inference on `chain`.

**Reply by Raynos** [reply]
The first thing I can mention is that typescript does not inline `interface Foo` like it does `type Foo = ...`

So for my own sanity I did `interface X extends ViemClient` to wrap it up into a box.

**Reply by Raynos** [reply]
I don't think this is an easy fix, i'd have to refactor many layers to come up with a prootytpe.

**Reply by jxom** [reply]
The large inline is a side-effect of the [`Prettify` type utility](https://github.com/wagmi-dev/viem/blob/main/src/types/utils.ts#L197-L207) – I don't think changing to an `interface` would have much of an effect... See here: https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzjAnmApnAKlAhgOwGdcBjGYCfOAXzgDMoIQ4ByAN2HRBYCgfUMcAApR0McnRQAeLAD44AXkQ84cANoBpOMCoBrdCgh1sAXQBc2TSZ60AZImp8SlQvBh4ipcpQD6ARgscAmIyCiolfABXABtongB6eNVVLDRMFijYlm1COHwIeFxCQmAAc3xcACNozBgIZDTWIM9QyhYAOgAKACYAZm7ugEonFzcPEO98H26LETEJaWaJsPkImLjE5OxGjPXs4Fz8wuKyiura+oF0pGqIEl0ACSKACwsAAwAGAA8AEgRXKA6UrUN5wAA+eXWAG44Ld7gA5SIgSroKAWSplHTwCGZaIwhhMd7fP4AoEgmGlIrozH4GAw54vIm-f7uMlvGE6MCRGBMkms-DA9l5SgkdAWKLI1EwtFwT7M0kC8lwQi8lmAxVCuqqhWC8GQ2Iw9zBLxhACS+AAJugvuKkSioHrcTD2i64AAOOAgaCYF3tQ1pCwAIhqlJIKEDMNoEKQvsjepjLsjHR6-SGQA

Maybe we could remove this for some types where it may not make much sense (for types that are heavily inferred).
